pywbem_mock/_defaultinstanceprovider.py

```diff
@@ -0,0 +1,439 @@
+    """
+    # TODO: This method modifies instance.
+    for iprop_name in instance:
+        if iprop_name not in target_class.properties:
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("Property {0!A} specified in NewInstance is not "
+                        "exposed by class {1!A} in namespace {2!A}",
+                        iprop_name, target_class.classname, namespace))
+
+        cprop = target_class.properties[iprop_name]
+        iprop = instance.properties[iprop_name]
+        if iprop.is_array != cprop.is_array or \
+                iprop.type != cprop.type:
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("Instance and class property {0!A} types "
+                        "do not match: instance={1!A}, class={2!A}",
+                        iprop_name, iprop, cprop))
+
+        # The class and instnames are the same except for possible case
+        # sensitivity. If case different, set cprop_name into new instance
+        # to maintain case equality
+        cprop_name = cprop.name
+        if cprop_name != iprop_name:
+            instance.properties[iprop_name].name = cprop_name
+
+        # If property not in instance, add it from class and use default value
+        # from class
+        for cprop_name in target_class.properties:
+            if cprop_name not in instance:
+                default_value = target_class.properties[cprop_name]
+                instance[cprop_name] = default_value
+
+
+class ProviderDispatcher(BaseProvider):
+    """
+    This class dispatches requests destined for the instance provider methods
+    defined in DefaultInstanceProvider  to the either the default instance
+    provder method or the provider defined for processing the defined class in
+    the defined namespace.
+    """
+    def __init__(self, conn, default_instance_provider):
+        """
+        """
+        # TODO: This not correct way to make cimrepository available to
+        # multiple subclasses
+        super(ProviderDispatcher, self).__init__(
+            conn=None, cimrepository=None)
+        self.default_instance_provider = default_instance_provider
+
+        self.cimrepository = conn.cimrepository
+        self.provider_registry = conn.provider_registry
+
+    def get_registered_provider(self, namespace, provider_type, classname):
+        """
+        If there is a provider registered for this namespace and classname
+        return that object (the instance of the DefaultInstanceProvider
+        subclass).
+
+        If no provider is registered, return None
+        """
+        if self.provider_registry:
+            if classname not in self.provider_registry:
+                return None
+            if self.provider_registry[classname]['providertype'] != provider_type:
+                return None
+            if not self.provider_registry[classname]['namespaces']:
+                return None
```

Comment:
Not sure about this. Are you allowing namespaces = None to indicate that it is registered for all namespaces? In that case, it should return the provider.

Reply:
After your proposal to allow different providers for different namespaces, I think that the combination of that and the idea of all namespaces are largely incompatible since we how have to allow the namespace to be specifically defined as a dictionary entry in the provider_registry. Therefore, I propose to drop that as a requirement and force all desired namespaces to be defined.

TODO: Not done yet

Reply:
Ok